debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbStreamingChangeEventSource.java

```diff
@@ -0,0 +1,406 @@
+
+import com.mongodb.CursorType;
+import com.mongodb.MongoClient;
+import com.mongodb.ServerAddress;
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.model.Filters;
+
+import io.debezium.connector.mongodb.ConnectionContext.MongoPrimary;
+import io.debezium.pipeline.ErrorHandler;
+import io.debezium.pipeline.EventDispatcher;
+import io.debezium.pipeline.source.spi.StreamingChangeEventSource;
+import io.debezium.pipeline.txmetadata.TransactionContext;
+import io.debezium.util.Clock;
+import io.debezium.util.Metronome;
+import io.debezium.util.Threads;
+
+/**
+ *
+ * @author Chris Cranford
+ */
+public class MongoDbStreamingChangeEventSource implements StreamingChangeEventSource {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbStreamingChangeEventSource.class);
+
+    private static final String AUTHORIZATION_FAILURE_MESSAGE = "Command failed with error 13";
+
+    private static final String OPERATION_FIELD = "op";
+    private static final String OBJECT_FIELD = "o";
+    private static final String OPERATION_CONTROL = "c";
+    private static final String TX_OPS = "applyOps";
+
+    private final EventDispatcher<CollectionId> dispatcher;
+    private final ErrorHandler errorHandler;
+    private final Clock clock;
+    private final MongoDbOffsetContext offsetContext;
+    private final ConnectionContext connectionContext;
+    private final ReplicaSets replicaSets;
+    private final MongoDbTaskContext taskContext;
+
+    public MongoDbStreamingChangeEventSource(MongoDbConnectorConfig connectorConfig, MongoDbTaskContext taskContext,
+                                             ReplicaSets replicaSets, MongoDbOffsetContext offsetContext,
+                                             EventDispatcher<CollectionId> dispatcher, ErrorHandler errorHandler, Clock clock) {
+        this.connectionContext = taskContext.getConnectionContext();
+        this.dispatcher = dispatcher;
+        this.errorHandler = errorHandler;
+        this.clock = clock;
+        this.replicaSets = replicaSets;
+        this.taskContext = taskContext;
+        this.offsetContext = (offsetContext != null) ? offsetContext : initializeOffsets(connectorConfig, replicaSets);
+    }
+
+    @Override
+    public void execute(ChangeEventSourceContext context) throws InterruptedException {
+        // Starts a thread for each replica-set and executes the streaming process
+        final int threads = replicaSets.replicaSetCount();
+        final ExecutorService executor = Threads.newFixedThreadPool(MongoDbConnector.class, taskContext.serverName(), "replicator-streaming", threads);
+        final CountDownLatch latch = new CountDownLatch(threads);
+
+        LOGGER.info("Starting {} thread(s) to stream changes for replica sets: {}", threads, replicaSets);
+        replicaSets.validReplicaSets().forEach(replicaSet -> {
+            executor.submit(() -> {
```

Comment:
Do we need a separate thread here?

Reply:
The task submitted to the executor does block when executing the `read from oplog` call using the `primaryClient.execute` call.  Since a connector may be configured with multiple replica-sets, I believe this has to remain for now. We might be able to work around this a bit when migrating to change streams, but that's a tbd.

Reply:
@Naros I belive we could do it after the release. It seems to me the only blocking operation is `com.mongodb.client.MongoCursor.next()` which can be replaced with `com.mongodb.client.MongoCursor.tryNext()`. But taking multiple rs into considertion this is definitely larger task and requires further discussion.

Reply:
I've added this as one todo item to [DBZ-1860](https://issues.redhat.com/browse/DBZ-1860).